lib/BackbeatConsumer.js

```diff
@@ -922,14 +920,28 @@ class BackbeatConsumer extends EventEmitter {
             next => {
                 if (this._zookeeper) {
                     this._zookeeper.close();
+                } else {
+                    process.nextTick(next);
                 }
-                if (this._consumer) {
+            },
+            next => {
+                const subscription = this._consumer.subscription() || [];
+                if (subscription.length > 0) {
                     this._consumer.unsubscribe();
-                    this._consumer.disconnect();
-                    this._consumer.on('disconnected', () => next());
+                    // Wait for partition unassign to complete before
+                    // disconnecting, the rebalance callback will handle
+                    // waiting for current jobs to complete as well as commit
+                    // the latest offsets
+                    this.on('unassign', next);
                 } else {
-                    process.nextTick(next);
+                    next(UNASSIGN_STATUS.IDLE);
+                }
+            },
+            (status, next) => {
+                if (status !== UNASSIGN_STATUS.TIMEOUT) {
+                    this._consumer.disconnect();
                 }
+                this._consumer.once('disconnected', () => next());
```

Comment:
to be on the safe side, we should check if we are already disconnected.
(`consumer.isConnected()` only looks at the state of the "javascript" consumer, not the underlying librdkafka client ; and that state is updated on `disconnected` event)